rfcs/20200612-stream-executor-c-api.md

```diff
@@ -0,0 +1,473 @@
+
+[StreamExecutor](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/stream_executor/stream_executor_pimpl.h;l=73) is TensorFlow's main device manager, responsible for work execution and memory management. It provides a set of methods
+(such as
+[Memcpy](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/stream_executor/stream_executor_internal.h;l=240))
+that can be customized for a particular device.
+
+We propose a C API wrapper of a subset of methods in
+[StreamExecutorInterface](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/stream_executor/stream_executor_internal.h;l=166?q=StreamExecutorinterface)
+as an ABI-stable way to register a custom StreamExecutor platform.
+
+# User Benefits
+
+A decoupled way to add a new device to TensorFlow.
+
+*   Simpler process: Does not have to add a new build toolchain to TensorFlow
+*   Faster time-to-solution: Does not need code review from the TensorFlow team.
+*   Lower maintenance efforts: Only C-API-related changes could break the
+    integration. Unrelated TensorFlow changes would not break the code.
+       *    The C APIs may be changed during the initial experimental phase based 
+            on developer experience and feedback. When the APIs become more mature, 
+            we will try to keep them stable (in a best-effort manner) until the new 
+            TensorFlow stack is available.
+
+# Design Proposal
+
+## StreamExecutorInterface
+
+[StreamExecutorInterface](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/stream_executor/stream_executor_internal.h;l=166?q=StreamExecutorinterface)
+is quite large and some of its methods are only sporadically used. Therefore, we
+plan to wrap only a subset of key StreamExecutorInterface functionality.
+
+See proposed C API below:
+
+```cpp
+typedef SE_Stream_st* SE_Stream;
+typedef SE_Event_st* SE_Event;
+typedef SE_Timer_st* SE_Timer;
+typedef TF_Status* (*TF_StatusCallbackFn)(void*);
+
+typedef struct SE_PlatformId {
+ void* id;  // aka stream_executor::Platform::Id
+} SE_PlatformId;
+
+typedef struct SE_TimerFns {
+ const size_t struct_size;
+ uint64_t (*nanoseconds)(SE_Timer timer);
+ uint64_t (*microseconds)(SE_Timer timer);
+} SE_Timer;
+
+typedef struct SE_AllocatorStats {
+  const size_t struct_size;
+  int64_t num_allocs;
+  int64_t bytes_in_use;
+  int64_t peak_bytes_in_use;
+  int64_t largest_alloc_size;
+
+  bool has_bytes_limit;
+  int64_t bytes_limit;
+
+  int64_t bytes_reserved;
+  int64_t peak_bytes_reserved;
+
+  bool has_bytes_reservable_limit;
+  int64_t bytes_reservable_limit;
+
+  int64_t largest_free_block_bytes;
+} SE_AllocatorStats;
+
+typedef enum SE_EventStatus {
+  SE_EVENT_UNKNOWN,
+  SE_EVENT_ERROR,
+  SE_EVENT_PENDING,
+  SE_EVENT_COMPLETE,
+} SE_EventStatus;
+
+typedef struct SE_Options {
+  const size_t struct_size;
+  int ordinal;
```

Comment:
Perhaps use int32_t instead of int? There might be other similar cases below as well.

Reply:
Fixed by @yisitu's commit.